R/rdeephaven/R/table_ops.R

```diff
@@ -0,0 +1,595 @@
+setGeneric(
+  "merge",
+  function(x, y = NULL, ...) {
+    standardGeneric("merge")
+  }
+)
+
+# We need many implementations of 'merge' because of the kind of args it should support
+
+base_merge <- function(x, y, ...) {
+  arg_list <- c(unlist(x), unlist(y), unlist(c(...)))
+  if (length(arg_list) == 1) {
+    return(arg_list[[1]])
+  }
+  unwrapped_arg_list <- lapply(arg_list, strip_s4_wrapping)
+  return(new("TableHandle", .internal_rcpp_object = unwrapped_arg_list[[1]]$merge(unwrapped_arg_list[2:length(unwrapped_arg_list)])))
+}
+
+# supports merge(c(t1, t2, t3))
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "list", y = "missing"),
+  function(x, y = NULL) {
+    verify_type("x", x, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y))
+  }
+)
+
+# supports merge(t1) edge case
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "missing"),
+  function(x, y = NULL) {
+    return(base_merge(x, y))
+  }
+)
+
+# supports merge(t1, t2, t3)
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "TableHandle"),
+  function(x, y, ...) {
+    if (length(c(...)) != 0) {
+      verify_type("...", c(...), "TableHandle", "Deephaven TableHandle", FALSE)
+    }
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+# supports merge(t1, c(t2, t3))
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "list"),
+  function(x, y) {
+    verify_type("y", y, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y))
+  }
+)
+
```

Comment:
I still have some questions about this `merge`.

1. How does R handle dispatch for NULL?  Is NULL any type or is NULL a specific type?  If it is any type, how is disambiguation handled?  If it is a specific type, then it is a type we need to handle.

2. We have a nasty product of types that need to be handled.
Types are {TableHandle, list, NULL(maybe), missing}.  This would mean we would need to support:
* TH, missing
* list, missing
* NULL, missing (?)
* TH, TH
* list, list
* TH, list
* list, TH 
* NULL, TH (?)
* TH, NULL (?)
* NULL, list (?)
* list, NULL (?)
* NULL, NULL (?)

3. Can we simply handle the above with a union type so that we don't need to implement the cartesian product?